File: src/tools/interpolation/average.ts
import { defaultGetInterpolateCurrentValue } from '.';
import { lRGBToGamma, sRGBToLinear } from '../..';

import type { GetInterpolateValue } from '.';
import type { MValue, Properties, RGBA, VectorPoint } from '../..';

/**
 * # Average Neighbor Interpolation
 *
 * ## Description
 * Finds the avarage point in the reference data to the given point and returns its value.
 *
 * ## Usage
 * ```ts
 * import { averageInterpolation, PointIndexFast } from 'gis-tools-ts';
 * import type { VectorPoint } from 'gis-tools-ts';
 *
 * // We have m-value data that we want to interpolate
 * interface TempData { temp: number; }
 *
 * const pointIndex = new PointIndexFast<TempData>();
 * // add lots of points
 * pointIndex.insertLonLat(lon, lat, data);
 * // ....
 *
 * // given a point we are interested in
 * const point: VectorPoint = { x: 20, y: -40 };
 * //  get a collection of points relative to the point
 * const data = await pointIndex.searchRadius(point.x, point.y, radius);
 *
 * // interpolate
 * const interpolatedValue = averageInterpolation<TempData>(point, data, (p) => p.m.temp);
 * ```
 * @param _point - Point to interpolate around. Unused for this method
 * @param refData - Reference data to search from
 * @param getValue - Function to get value from reference data
 * defaults to function that returns the z value or 0 if the z value is undefined
 * @returns - The avarage value of the collection of points
 */
export function averageInterpolation<T extends MValue = Properties>(
  _point: VectorPoint,
  refData: VectorPoint<T>[],
  getValue: GetInterpolateValue<T> = defaultGetInterpolateCurrentValue,
): number {
  let total = 0;
  for (const refPoint of refData) total += getValue(refPoint);

  return total / refData.length;
}

/**
 * Helper function for {@link averageInterpolation} on RGB(A) data.
 * Light in RGB data is logarithmically weighted (gamma corrected), so we need to expand each component by n^2 to
 * get the correct weight for each component.
 * @param point - Point to interpolate
 * @param refData - Reference data points
 * @returns - The interpolated RGBA data.
 */
export function rgbaAverageInterpolation(point: VectorPoint, refData: VectorPoint<RGBA>[]): RGBA {
  if (refData.length === 0) return { r: 0, g: 0, b: 0, a: 255 };

  const rData = averageInterpolation(point, refData, (p) => sRGBToLinear(p.m?.r ?? 0));
  const gData = averageInterpolation(point, refData, (p) => sRGBToLinear(p.m?.g ?? 0));
  const bData = averageInterpolation(point, refData, (p) => sRGBToLinear(p.m?.b ?? 0));
  const a = averageInterpolation(point, refData, (p) => p.m?.a ?? 255); // Alpha stays linear!

  return {
    r: lRGBToGamma(rData),
    g: lRGBToGamma(gData),
    b: lRGBToGamma(bData),
    a, // No gamma correction on alpha
  };
}

File: rust/tools/interpolation/average.rs
use s2json::MValueCompatible;
pub use s2json::{MValue, ValueType, VectorPoint};

use super::{get_channel, RgbaChannel};
use crate::readers::RGBA;
use crate::tools::{default_get_interpolate_current_value, GetInterpolateValue, VectorPointRGBA};

/// # Average Neighbor Interpolation
///
/// ## Description
/// Finds the avarage point in the reference data to the given point and returns its value.
///
/// ## Usage
pub fn average_interpolation<T: MValueCompatible>(
    _point: &VectorPoint,
    ref_data: &[VectorPoint<T>],
    get_value: Option<GetInterpolateValue<T>>,
) -> f64 {
    let get_value = get_value.unwrap_or(default_get_interpolate_current_value);
    let mut total = 0.;
    for ref_point in ref_data {
        total += get_value(ref_point);
    }

    total / ref_data.len() as f64
}

/// Helper function for {@link averageInterpolation} on RGB(A) data.
/// Light in RGB data is logarithmically weighted, so we need to expand each component by n^2 to
/// get the correct weight for each component.
pub fn average_interpolation_rgba(point: &VectorPoint, ref_data: &[VectorPointRGBA]) -> RGBA {
    if ref_data.is_empty() {
        return RGBA::default();
    }
    let r = average_interpolation(point, ref_data, Some(|p| get_channel(p, RgbaChannel::R)));
    let g = average_interpolation(point, ref_data, Some(|p| get_channel(p, RgbaChannel::G)));
    let b = average_interpolation(point, ref_data, Some(|p| get_channel(p, RgbaChannel::B)));
    let a = average_interpolation(point, ref_data, Some(|p| get_channel(p, RgbaChannel::A)));

    RGBA::new(r, g, b, a)
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::{vec, vec::Vec};

    #[test]
    fn test_average_interpolation() {
        let point = VectorPoint::new(0.5, 0.5, None, None);
        let ref_data: Vec<VectorPoint> = vec![
            VectorPoint::new(0., 0., Some(1.), None),
            VectorPoint::new(1., 0., Some(2.), None),
            VectorPoint::new(0., 1., Some(3.), None),
            VectorPoint::new(1., 1., Some(4.), None),
        ];
        let result = average_interpolation(&point, &ref_data, None);
        assert_eq!(result, 2.5);
    }

    #[test]
    fn test_average_interpolation_rgba() {
        let ref_data: vec::Vec<VectorPointRGBA> = vec![
            VectorPointRGBA::new(0., 0., None, Some(RGBA::from_u8s(20, 20, 60, 255))),
            VectorPoint::new(1., 0., None, Some(RGBA::from_u8s(30, 100, 60, 255))),
            VectorPoint::new(0., 1., None, Some(RGBA::from_u8s(127, 127, 60, 255))),
            VectorPoint::new(1., 1., None, Some(RGBA::from_u8s(255, 255, 60, 255))),
        ];

        // test 1
        let point = VectorPoint::new(0.5, 0.5, None, None);
        let result = average_interpolation_rgba(&point, &ref_data);
        assert_eq!(result.to_u8s(), (84, 107, 60, 255));

        // test 2
        let point = VectorPoint::new(0.65, 0.15, None, None);
        let result = average_interpolation_rgba(&point, &ref_data);
        assert_eq!(result.to_u8s(), (84, 107, 60, 255));

        // test 3
        let result = average_interpolation_rgba(&point, &[]);
        assert_eq!(result.to_u8s(), (0, 0, 0, 255));
    }
}
